File: visitor/java/VisitorTest.java
import java.util.ArrayList;
import java.util.List;

/**
 * 多态是一种「动态绑定」，可以在运行时获取对象的实际类型，来运行实际类型对应的方法。
 * 而函数重载是一种「静态绑定」，在编译时并不能获取对象的实际类型，而是根据声明类型执行声明类型对应的方法。
 */
public class VisitorTest {
    public static void main(String[] args) {
        List<ResourceFile> resourceFiles = listAllResourceFiles("~/");

        Extractor extractor = new Extractor();
        for (ResourceFile resourceFile : resourceFiles) {
            // extractor.visit(resourceFile);
            resourceFile.accept(extractor);
        }

        System.out.println("==========================================");

        Compressor compressor = new Compressor();
        for (ResourceFile resourceFile : resourceFiles) {
            // compressor.visit(resourceFile);
            resourceFile.accept(compressor);
        }
    }

    private static List<ResourceFile> listAllResourceFiles(@SuppressWarnings("SameParameterValue")
                                                           String resourceDirectory) {
        List<ResourceFile> resourceFiles = new ArrayList<>();
        resourceFiles.add(new PdfFile(resourceDirectory + "a.pdf"));
        resourceFiles.add(new WordFile(resourceDirectory + "b.docx"));
        resourceFiles.add(new PPTFile(resourceDirectory + "c.ppt"));
        return resourceFiles;
    }
}

File: visitor/go/visitor_test.go
package visitor

import (
	"fmt"
	"testing"
)

func TestVisitor(t *testing.T) {
	resourceFiles := listAllResourceFiles("~/")

	extractor := &Extractor{}
	for _, f := range resourceFiles {
		// switch f.(type) ... ==> extractor.[VistPPTFile(f) | VistPdfFile(f) | VistWordFile(f)]
		f.Accept(extractor)
	}

	fmt.Println("==========================================")

	compressor := &Compressor{}
	for _, f := range resourceFiles {
		// switch f.(type) ... ==> compressor.[VistPPTFile(f) | VistPdfFile(f) | VistWordFile(f)]
		f.Accept(compressor)
	}
}

func listAllResourceFiles(resourceDirectory string) []ResourceFile {
	return []ResourceFile{
		NewPdfFile(resourceDirectory + "a.pdf"),
		NewWordFile(resourceDirectory + "b.docx"),
		NewPPTFile(resourceDirectory + "c.ppt"),
	}
}
